### PR TITLE
Issue 36 fix

### PR DESCRIPTION
**Pull request by JP-SystemsX**
Addressing Issue #36 

The _get_param function of the finetuning arguments should now:
- Retrieve All trainable Parameters as long as naming convention is kept
- Avoid Exploding learning rate due to bug in calculation of learning rate during Layer wise decreasing Learning rate
- Raise exception if a model doesn't keep the naming convention

Still has problems with:
- Models that don't keep naming conventions like ALBERT

### CODE REVIEW

File: small_text/integrations/transformers/classifiers/classification.py [hunk 1]
```diff
@@ -122,32 +122,55 @@ def __init__(self,
...
+        layers.append(model.classifier.parameters())
 
     total_layers = len(layers)
 
     use_gradual_unfreezing = isinstance(fine_tuning_arguments.gradual_unfreezing, int) and \
-        fine_tuning_arguments.gradual_unfreezing > 0
+                             fine_tuning_arguments.gradual_unfreezing > 0
 
-    start_layer = 0 if not use_gradual_unfreezing else total_layers-fine_tuning_arguments.gradual_unfreezing
+    start_layer = 0 if not use_gradual_unfreezing else max(0, total_layers - fine_tuning_arguments.gradual_unfreezing)
```
Comment (chschroeder) [inline small_text/integrations/transformers/classifiers/classification.py:152]: Why the `max()`? I think if we check that the number of gradual unfreezing parameters is less or equal than the number of layers, which we should do anyway, this should not be needed.

**Review by chschroeder — CHANGES_REQUESTED** [review]
Thanks for your effort! I like the solution in general. It has less bugs than the version before, but still some points that could make problems.

I would like to get this as reliable as possible, so that we don't end up with a similar bug in the future. I hope you can bear with me on this one ;).

What I was thinking was more in the direction of the following:
```
def _get_layer_params(model: PreTrainedModel, base_lr: float, fine_tuning_arguments: FineTuningArguments):

    base_model = getattr(model, model.base_model_prefix)
    params_dict, encoder_name = _build_parameter_dict(model, base_model)

    total_layers = len(params_dict)

    layerwise_gradient_decay = fine_tuning_arguments.layerwise_gradient_decay
    use_gradual_unfreezing = isinstance(fine_tuning_arguments.gradual_unfreezing, int) and \
        fine_tuning_arguments.gradual_unfreezing > 0

    params = [{'params': param, 'lr': base_lr } for _, param in params_dict.items()]

    if use_gradual_unfreezing:

        start_layer = 0 if not use_gradual_unfreezing else total_layers-fine_tuning_arguments.gradual_unfreezing
        num_layers = total_layers - start_layer

        for i in range(start_layer, total_layers):
            lr = base_lr if not layerwise_gradient_decay else base_lr * layerwise_gradient_decay ** (
                        num_layers - i)
            params[i]['lr'] = lr

    return params


def _build_parameter_dict(model, base_model):
    params_dict = {}

    for name, param in base_model.named_children():
        params_dict[name] = param

    encoder_name = _identify_and_remove_encoder_block(params_dict)

    # TODO: assert that 'layer' exists
    encoder_layers = dict(getattr(base_model,'encoder').named_children())['layer']
    for name, param in encoder_layers.named_children():
        params_dict[name] = param

    for name, param in model.named_children():
        if name != model.base_model_prefix:
            params_dict[name] = param

    return params_dict, encoder_name


def _identify_and_remove_encoder_block(params_dict):
    module_names = list(params_dict.keys())
    last_module_name = module_names[-1]

    if last_module_name != 'encoder' and last_module_name != 'transformer':
        raise ValueError('Encoder could not be recognized. This model is not supported.')

    del params_dict[last_module_name]

    return last_module_name
```

What do you think about this? Is it more readable?

File: small_text/integrations/transformers/classifiers/classification.py [hunk 2]
```diff
@@ -153,8 +153,7 @@ def _get_layer_params(model, base_lr, fine_tuning_arguments):
     num_layers = total_layers - start_layer
 
     for i in range(start_layer, total_layers):
-        lr = base_lr if not layerwise_gradient_decay else base_lr * layerwise_gradient_decay ** (
-                num_layers - i)
+        lr = base_lr if not layerwise_gradient_decay else base_lr * layerwise_gradient_decay ** (num_layers - (i+1-start_layer))
```
Comment (chschroeder) [inline small_text/integrations/transformers/classifiers/classification.py:156]: Great you caught this! The problem here is a mix up between `num_layers` and `total_layers`.

File: small_text/integrations/transformers/classifiers/classification.py (hunk 1 above)
Comment (JP-SystemsX) [reply small_text/integrations/transformers/classifiers/classification.py:152]: Certainly, I just didn't saw where this was done, if it was done somewhere

File: small_text/integrations/transformers/classifiers/classification.py [hunk 3]
```diff
@@ -122,32 +122,55 @@ def __init__(self,
...
 
     for i in range(start_layer, total_layers):
         lr = base_lr if not layerwise_gradient_decay else base_lr * layerwise_gradient_decay ** (
-                    num_layers - i)
-        params.append({
-            'params': layers[i].parameters(),
-            'lr': lr
-        })
+                num_layers - i)
+        for sublayer in layers[i]:
```
[the remark this answers is not in the document]
Comment (JP-SystemsX) [reply small_text/integrations/transformers/classifiers/classification.py:157]: This is just a neat little way to get the layers such that one can make the check at the end via the Ids more easily dont know if this can be done easier via pytorch. 
Bonus: one can remove the layers that were frozen by the pytorch interface immediately
And yes it probably should be done recursively and be better documented.

File: small_text/integrations/transformers/classifiers/classification.py [hunk 4]
```diff
@@ -122,32 +122,55 @@ def __init__(self,
...
 def _get_layer_params(model, base_lr, fine_tuning_arguments):
-
     layerwise_gradient_decay = fine_tuning_arguments.layerwise_gradient_decay
 
     params = []
 
+    # Get layers under assumption that a certain naming convention is kept
     base_model = getattr(model, model.base_model_prefix)
+    layers = []
+    if hasattr(base_model, 'embeddings'):
```
[the remark this answers is not in the document]
Comment (JP-SystemsX) [reply small_text/integrations/transformers/classifiers/classification.py:132]: Yeah I also thought of using named_children but the problem is here that, to the best of my knowledge, they are given in the ordering as they are defined in the model not in the one how they are connected, right? Which, in my mind at least, can be even more arbitrary than the naming convention (because no one will blame you if you defined the encoder after the classifier as long as you connected them the right way, right? But if one named them weirdly they might get frowned upon) therefore one might need to redefine the loop in something that uses the names again which would bring us back to the start?

Plus: Changing the names is often done with great care because that makes versions easily incompatible with each other, reordering on the other hand can be done super easily and can therefore change quite swiftly

File: small_text/integrations/transformers/classifiers/classification.py (hunk 2 above)
Comment (JP-SystemsX) [reply small_text/integrations/transformers/classifiers/classification.py:156]: Ah that explains it, thanks.

File: small_text/integrations/transformers/classifiers/classification.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chschroeder) [reply small_text/integrations/transformers/classifiers/classification.py:132]: Ugh, good point. This leaves us with very few options for a generic mechanism:
1. We assume a fixed setup where certain modules have to be present and whose names have to be what we expect them to be.
2. We provide this per model (which would be a lot of effort).

In this case, I have to go back and support your solution (1.) to check for fixed names.

**Review by chschroeder — APPROVED** [review]

**Comment by chschroeder** [comment]
After the discussion, I think this solution is the right way. Thanks!

Do you have any pending tasks here? Otherwise I would just merge it. 

After that I would like to add 1-2 unit tests (which were apparently lacking in this area of the code), and afterwards it should be done.

**Comment by chschroeder** [comment]
@JP-SystemsX See the comment above regarding your question in #36 on how to proceed.
